core/tx_list.go

```diff
@@ -360,6 +360,30 @@ func (l *txList) Filter(costLimit *big.Int, gasLimit uint64) (types.Transactions
 	return removed, invalids
 }
 
+// FilterF removes all transactions from the list that satisfy a predicate.
+// Every removed transaction is returned for any post-removal maintenance.
+// Strict-mode invalidated transactions are also returned.
+func (l *txList) FilterF(f func(tx *types.Transaction) bool) (types.Transactions, types.Transactions) {
```

Comment:
What will be the effect of removing this part?

```go
// If all transactions are below the threshold, short circuit
if l.costcap.Cmp(costLimit) <= 0 && l.gascap <= gasLimit {
	return nil, nil
}
l.costcap = new(big.Int).Set(costLimit) // Lower the caps to the thresholds
l.gascap = gasLimit
```

Reply:
I dont know, so I am not inclined to remove it.

Reply:
so that it needs to be added to this function?

Reply:
The decision is non-trivial. If we add the early return check, we won't check L1 data fee for those transactions. If we don't, then txpool will need to do much more work.

As for setting / not setting these two, I'm not sure about the impact:

```go
l.costcap = new(big.Int).Set(costLimit) // Lower the caps to the thresholds
l.gascap = gasLimit
```

Reply:
yeah. I also noticed that the behavior of this heuristic changed (as that said, non-trivial) after considering the L1 data fee.
adding compatibility in this check requires changing the definition of `l.costcap` (adding L1 data fee):
```
l.txs.Put(tx)
if cost := tx.Cost(); l.costcap.Cmp(cost) < 0 {
	l.costcap = cost
}
```

Reply:
alright, lets put it back.

I dont recall why i removed it, maybe it was by accident.

Reply:
> If we add the early return check, we won't check L1 data fee for those transactions

maybe this is why?

Reply:
ah that makes sense

Reply:
We can consider adding the L1 data fee in `l.costcap`, thus the logic won't break anything.

Reply:
added in [8d383fb](https://github.com/scroll-tech/go-ethereum/pull/627/commits/8d383fb6b7a04e63cd34a1d08da0b9e5528cb601).